facilitator/src/bin/facilitator.rs

```diff
@@ -1295,3 +1308,90 @@ fn transport_for_path(
         StoragePath::LocalPath(path) => Ok(Box::new(LocalFileTransport::new(path))),
     }
 }
+
+fn decode_base64_key(s: &str) -> Result<Vec<u8>> {
+    if s == "not-a-real-key" {
+        return Err(anyhow!(
+            "'not-a-real-key'. Run deploy-tool to generate secrets"
+        ));
+    }
+    base64::decode(s).context("decoding key from base64")
+}
+
+// You can't make a trait object out of a trait that is generic in another trait
+// (as would be the case for TaskQueue<T: Task>) because such traits are not
+// "object safe" [1], so we can't write a function like
+// fn task_queue_from_args<T: Task>() -> Result<Box<dyn TaskQueue<T>>>.
+// To work around this we manually provide specializations on
+// task_queue_from_args for IntakeBatchTask and AggregationTask.
```

Comment:
Since you're providing specializations, do you really need the trait object? Seems like you could return `Result<Box<TaskQueue<IntakeBatchTask>>>` for this one, and `Result<Box<TaskQueue<AggregateTask>>>` for the other.

I think you could collapse these two into a single implementation that way:

```rust
fn task_queue_from_args<T>(matches: &ArgMatches) -> Result<Box<TaskQueue<T>>>
```

Reply:
I don't think I understand your comment. I _do_ return `Result<Box<dyn TaskQueue<IntakeBatchTask>>>` from the one and `Result<Box<dyn TaskQueue<AggregationTask>>>` from the other. And I tried to write almost exactly the function signature you suggest, though with `Result<Box<dyn TaskQueue<T>>>` and `T: Task`. You can't do it, because `TaskQueue` is not an object safe trait. I'm pretty dumb, though, so maybe there's some easy way to do it that I missed.

Reply:
I'm saying: I don't think you need the `dyn` (which makes it a trait object). And actually you may not need the `Box` either, if it was only there for the purpose of making a trait object.

Reply:
Oops, sorry, I realized my mistake here. What makes this a trait object is that `TaskQueue` is a trait, not the `dyn`. The `dyn` is just syntactic sugar (gradually becoming required) to emphasize to the reader that `TaskQueue` is a trait and therefore we're using a trait object here.

So, code is fine as it stands and now I understand what you were getting at. Sorry for the misleading / confusing comments!

Reply:
No worries, this is actually quite confusing! I also tried to ameliorate this by changing `TaskHandle` to not have any type parameters, and use `Task` as a trait object, but that didn't work out because then you can't get a trait object from a `Sized`, and if you're not `Sized`, you can't implement `serde::de::DeserializedOwned`... kind of a mess really, so I'm happy to live with these two ugly functions.